plugins/cameraserver/src/main/java/edu/wpi/first/shuffleboard/plugin/cameraserver/source/CameraUrlGenerator.java

```diff
@@ -45,31 +51,95 @@ public CameraUrlGenerator(CameraServerSource source) {
+    }
+    return generateUrls(commands, baseUrls, source.getName());
   }
 
   @VisibleForTesting
-  static String[] generateUrls(Map<String, String> commands, String[] baseUrls) { // NOPMD varargs instead of array
+  static String[] generateUrls(Map<String, String> commands, String[] baseUrls,
+                               String cameraName) {
     if (baseUrls == null || baseUrls.length == 0) {
       return new String[0];
     }
     if (commands.isEmpty()) {
       return baseUrls;
     } else {
-      return Arrays.stream(baseUrls)
-          .map(url -> url + toHttpParams(commands))
+      var urls =  Arrays.stream(baseUrls)
+          .map(url -> addHttpParams(url, commands))
+          .filter(url -> url != null)
           .toArray(String[]::new);
+      return urls;
     }
   }
 
   @VisibleForTesting
-  static String toHttpParams(Map<String, String> commands) {
+  static String addHttpParams(String input, Map<String, String> commands) {
     if (commands.isEmpty()) {
-      return "";
+      return input;
+    }
+    // Parse the URI
+    URI uri;
+    try {
+      uri = new URI(input);
+    } catch (URISyntaxException ex) {
+      return input;
     }
-    return commands.entrySet().stream()
-        .map(e -> e.getKey() + "=" + e.getValue())
+
+    String query = uri.getQuery();
+
+    if (query != null) {
+      // Handle the NI special case
+      String niName = commands.getOrDefault("name", null);
+
+      String[] existingCommands = query.split("&");
+      for (String command : existingCommands) {
+        String[] commandSplit = command.split("=");
+        if (commandSplit.length != 2) {
+          continue;
+        }
+        try {
+          commands.put(URLDecoder.decode(commandSplit[0], "utf-8"), URLDecoder.decode(commandSplit[1], "utf-8"));
+        } catch (UnsupportedEncodingException e1) {
+          commands.put(commandSplit[0], commandSplit[1]);
+        }
+      }
+      if (niName != null) {
+        commands.put("name", niName);
+      }
+    }
+
+    var queryStr = commands.entrySet().stream()
+        .map(e -> httpUrlEncode(e))
         .collect(Collectors.joining("&"));
+      return encodeUri(uri, queryStr);
+
+  }
+
+  private static String encodeUri(URI uri, String queryStr) {
+    StringBuilder builder = new StringBuilder();
+    builder.append(uri.getScheme());
+    builder.append("://");
+    builder.append(uri.getAuthority());
+    builder.append(uri.getPath());
+    builder.append('?');
+    builder.append(queryStr);
+    String fragment = uri.getFragment();
+    if (fragment != null) {
+      builder.append('#');
+      builder.append(fragment);
+    }
+    return builder.toString();
   }
 
+  private static String httpUrlEncode(Map.Entry<String, String> rawCommand) {
+    try {
+      return URLEncoder.encode(rawCommand.getKey(), "utf-8").replaceAll("\\+", "%20") + "="
+          + URLEncoder.encode(rawCommand.getValue(), "utf-8").replaceAll("\\+", "%20") ;
```

Comment:
If you're just replacing the `+` character, use `replace("+", "%20")`. Faster and more obvious than a regular expression

Reply:
There can be multiple `+` characters, so `replace` doesn't work.

Reply:
`String.replace` replaces all instances of the target string: https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/String.html#replace(java.lang.CharSequence,java.lang.CharSequence)

Reply:
Hmm I wonder where I thought it didn't replace all. I think I ran into that when working with gradle or something. If thats what it does then that works. Fixed